text/2021-09-27-refine-error-handling.md

```diff
@@ -0,0 +1,293 @@
+}
+```
+
+### Error Wrapping
+
+* Return the original error if there is no additional context to add
+* Add context using
+  [`"pkg/errors".Wrap`](https://pkg.go.dev/github.com/pkg/errors#Wrap) so that
+  the error message provides more context
+* Use [`"pkg/errors".Errorf`](https://pkg.go.dev/github.com/pkg/errors#Errorf)
+  with if the callers do not need to detect or handle that specific error case.
+
+The [`"pkg/errors".Wrap`](https://pkg.go.dev/github.com/pkg/errors#Wrap) will
+add a stack trace, so you don't need to wrap it with `WithStack` again.
+
+The context usually includes: what you are doing, the object of the operation
+(like the pod name, the chaos name...).
+
+```go
+func startProcess(cmd *exec.Cmd) error {
+   err := cmd.Start()
+   if err != nil {
+      return nil, errors.Errorf("start process: %v", err)
+   }
+
+   return nil
+}
+```
+
+A more realistic example in `chaos-daemon` is:
+
+```go
+func (s *DaemonServer) ExecStressors(ctx context.Context,
+    req *pb.ExecStressRequest) (*pb.ExecStressResponse, error) {
+   ...
+
+   control, err := cgroups.Load(daemonCgroups.V1, daemonCgroups.PidPath(int(pid)))
+   if err != nil {
+      return nil, errors.Wrapf(err, "load cgroup of pid %d", pid)
+   }
+
+   ...
+}
+```
+
+When adding context to returned errors, keep the context succinct by avoiding
+phrases like "failed to", which state the obvious and pile up as the error
+percolates up through the stack:
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+s, err := store.New()
+if err != nil {
+   return fmt.Errorf(
+      "failed to create new store: %v", err)
+}
+```
+
+</td><td>
+
+```go
+s, err := store.New()
+if err != nil {
+   return fmt.Errorf(
+      "new store: %v", err)
+}
+```
+
+</td></tr>
+</tbody>
+</table>
+
+### Error Handling
+
+The key of error handling is to identify whether this error is an expected one.
+For example, if the Chaos Mesh is recovering an injection, but the kubernetes
+returns `PodNotFound`, we need to "handle" it by just ignoring.
+
+Identifying errors could have two different situation: if the target error is a
+type, you could use `errors.As` to extract the error information, and if the
+target error is a variable (created by `errors.New`, in most situation), you
+could use `errors.Is`.
+
+The `error` type of golang is really a chaos, a type and a varaible can both
+mean a kind of error, and you should treat them in different pattern. To make
+the thing simpler, you should **never** use `errors.Is` to identify a customized
+error type (other than the variable created by `errors.New`). It's prefered to
+use `errors.As` and then judge the equality (unless your error is recursively
+wrapped, which is not common, right?)
+
+#### Handling Error from Kubernetes
+
+The error from kubernetes client is usually derived from an HTTP status. The
+only way to identify them is to use `k8sError.Is*`, e.g.
+`k8sError.IsNotFound`.
+
+#### Handling Error from Grpc
```

Comment:
I think a more detailed advice about how to handle errors propagating from chaos-daemon to chaos-controller is necessary.

Reply:
Example added.